torch/csrc/api/src/nn/modules/embedding.cpp

```diff
@@ -11,27 +12,206 @@
 namespace torch {
 namespace nn {
 
-EmbeddingOptions::EmbeddingOptions(int64_t count, int64_t dimension)
-    : count_(count), dimension_(dimension) {}
+    EmbeddingImpl::EmbeddingImpl(const EmbeddingOptions& options) : options(options) {
+      reset();
+    }
 
-EmbeddingImpl::EmbeddingImpl(EmbeddingOptions options) : options(options) {
-  reset();
-}
+    void EmbeddingImpl::reset() {
+      if (options.padding_idx() != c10::nullopt) {
+        if(*options.padding_idx() > 0) {
+          TORCH_CHECK(*options.padding_idx() < options.num_embeddings(), "Padding_idx must be within num_embeddings");
+        }
+        else if(*options.padding_idx() < 0) {
+          TORCH_CHECK(*options.padding_idx() >= -(options.num_embeddings()), "Padding_idx must be within num_embedding");
+          options.padding_idx(options.num_embeddings() + *options.padding_idx());
+        }
+      }
 
-void EmbeddingImpl::reset() {
-  weight = register_parameter(
-      "weight", torch::empty({options.count(), options.dimension()}));
-  NoGradGuard guard;
-  weight.normal_(0, 1);
-}
+      if (options._weight() == c10::nullopt) {
+        weight = register_parameter(
+            "weight", torch::empty({options.num_embeddings(), options.embedding_dim()}));
+        torch::nn::init::normal_(weight);
+        if(options.padding_idx() != c10::nullopt) {
+          torch::NoGradGuard no_grad;
+          weight[*options.padding_idx()].fill_(0);
+        }
+      }
+      else {
+        TORCH_CHECK(((*options._weight()).size(0) == options.num_embeddings()) && ((*options._weight()).size(1) == options.embedding_dim()), "Shape of _weight does not match num_embeddings and embedding_dim");
```

Comment:
nit: we can write `(*options._weight()).sizes() == torch::IntArrayRef({options.num_embeddings(), options.embedding_dim()})` here, to better match Python version.